### PR TITLE
feat(auth): Prevent redirect when OAuth errors

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
## Objective:
It was discovered that the open redirect vulnerability exists in the “API Applications” feature, effectively opening the possibilities for an attacker to conduct Phishing attacks, or chain it with other bugs to elevate impact. The issue lies in the OAuth endpoint created by the “API Applications” feature. When an error occurs during the OAuth processing, the application redirects to the user-specified URL without any confirmation.

### Steps to reproduce:
1. Log-in
2. Navigate to https://sentry.io/settings/account/api/applications/
3. Click on the “Create New Application” button
4. Enter “https://example.com/” to the “Authorized Redirect URIs” field
5. Copy the “Client ID”
6. Concatenate the copied ID with “https://sentry.io/oauth/authorize/?client_id=”
7. Navigate to https://sentry.io/oauth/authorize/?client_id=[CLIENT_ID]. It will
redirect to https://example.com/ directly

While the issue itself has a low severity, open redirects in many cases operate as very useful gadgets for forming chains with other issues.

It is recommended to return a 400 error page rather than redirect to the URL specified in the “Authorized Redirect URIs” when an error happens in the OAuth processing.

So in this PR, I made Sentry redirect to the 400 error page. I also refactored to remove duplicated logic.

### CODE REVIEW

File: tests/sentry/web/frontend/test_oauth_authorize.py [hunk 1]
```diff
@@ -27,8 +27,9 @@ def test_missing_response_type(self):
             )
         )
 
-        assert resp.status_code == 302
-        assert resp["Location"] == "https://example.com?error=unsupported_response_type"
+        assert resp.status_code == 200
```
Comment (manuzope) [inline tests/sentry/web/frontend/test_oauth_authorize.py:30]: Any reason why this is a 200 and not a 400?

**Comment by NisanthanNanthakumar** [comment]
> Hi @NisanthanNanthakumar
> Out of curiosity, did you manage to validate how an attacker could manipulate the `Authorized Redirect URIs` parameter? From the steps you described, this sounds like a parameter only the app/service provider can alter.

Hi @colin-ife-snyk, this PR is for an issue in the OAuth specification outlined here: https://hackerone.com/reports/26962 .

**Comment by JTCunning** [comment]
Removed myself from code review since I am not to be trusted with reviewing more than three lines of Python.

**Review by scefali — APPROVED** [review]
LGTM

**Pull request merged**